view/newPerspective/utils.js

```diff
@@ -235,52 +235,94 @@ function getPublishedFromArr(arr) {
-    rootSubject: {},
-    lens: {}, // includes library
   };
+  let hierarchyLoadEvent; // will be custom event
+  let gotLens = false;
 
+  // get the perspectives first,
+  // in case url ends with /perspectives
+  // and we need to get the first perspective by alphabetical order
   return request('/v1/perspectives')
-  .then((res) => {
-    valuesObj.persNames = res.body.map((perspective) => perspective.name).sort();
+  .then((response) => {
+
+    // assign perspective-related values to the accumulator object
+    valuesObj.persNames = response.body.map((perspective) => perspective.name).sort();
+
+    // get the perspective name. If no perspective found, getPerspectiveName throws an error
+    // and the rest won't be executed.
     valuesObj.name = getPerspectiveName(valuesObj.persNames);
-    valuesObj.perspective = res.body.filter((perspective) => perspective.name === valuesObj.name)[0];
+    valuesObj.perspective = response.body.filter((perspective) => perspective.name === valuesObj.name)[0];
 
-    // need to get the lens, as GET /lenses does not return the library
-    return request('/v1/lenses/' + valuesObj.perspective.lensId);
-  })
-  .then((res) => {
-    valuesObj.lens = res.body;
-    return request('/v1/subjects?fields=isPublished,absolutePath,tags');
-  })
-  .then((res) => {
+    // Whenever we get the response back with the hierarchy, we dispatch the
+    // lens.hierarchyLoad event to the lens. (If we happen to get the hierarchy
+    // back *before* the lens, hold onto it, wait for the lens, *then* dispatch the
+    // lens.hierarchyLoad event *after* the lens.load event.)
+    const getLens = request('/v1/lenses/' + valuesObj.perspective.lensId)
+    .then((res) => {
+
+      // hierarchyLoadEvent can be undefined or a custom event
+      handleLensDomEvent(res.body, hierarchyLoadEvent);
+
+      // if hierarchy is not loaded, set the lens received flag
+      // to true, to dispatch lens load when hierarchy
+      // is returned in getHierarchy
+      gotLens = true;
+
+      return res;
+    });
 
-    // set the published subjects
-    valuesObj.subjects = getPublishedFromArr(res.body);
-    valuesObj.subjectTagFilter = getTagsFromArrays(valuesObj.subjects);
     const filterString  = getFilterQuery(valuesObj.perspective);
-    return request('/v1/subjects/' + valuesObj.perspective.rootSubject + '/hierarchy' + filterString);
-  })
-  .then((res) => {
-    valuesObj.rootSubject = res.body;
-    return request('/v1/lenses');
-  })
-  .then((res) => {
-    valuesObj.lenses = getPublishedFromArr(res.body);
-    return request('/v1/aspects?fields=isPublished,name,tags')
+    const getHierarchy = request('/v1/subjects/' +
+      valuesObj.perspective.rootSubject + '/hierarchy' + filterString)
+    .then((res) => {
+
+      // load the hierarchy event
+      handleHierarchyEvent(res.body, gotLens);
+
+      return res;
+    });
+
+    const promisesArr = [
+      getLens,
+      getHierarchy,
+      request('/v1/lenses'),
```

Comment:
refactored to get the perspectives, get the perspective, then get everything else.
Deployed onto devbox. Response time has decreased slightly from 3s to 2.2s
original: https://refocus-devbox.herokuapp.com/perspectives/NorthAmerica33
new: https://refocus-devbox.herokuapp.com/perspectives/NorthAmerica33?beta=true
![screen shot 2017-06-02 at 5 49 53 pm](https://cloud.githubusercontent.com/assets/17040750/26749308/dbc99334-47bc-11e7-85d1-d7fc886b42c8.png)
![screen shot 2017-06-02 at 5 49 45 pm](https://cloud.githubusercontent.com/assets/17040750/26749309/dbcaad50-47bc-11e7-85ce-fb14592d98ef.png)